services/acm/src/main/java/software/amazon/awssdk/services/acm/auth/scheme/internal/AcmAuthSchemeInterceptor.java

```diff
@@ -0,0 +1,133 @@
+
+    @Override
+    public void beforeExecution(Context.BeforeExecution context, ExecutionAttributes executionAttributes) {
+        List<AuthSchemeOption> authOptions = resolveAuthOptions(executionAttributes);
+        SelectedAuthScheme<?> selectedAuthScheme = selectAuthScheme(authOptions, executionAttributes);
+        executionAttributes.putAttribute(SdkInternalExecutionAttribute.SELECTED_AUTH_SCHEME, selectedAuthScheme);
+    }
+
+    /**
+     * Invoke the auth scheme resolver to determine which auth options we should consider for this request.
+     */
+    private List<AuthSchemeOption> resolveAuthOptions(ExecutionAttributes executionAttributes) {
+        // Prepare the inputs for the auth scheme resolver. We always include the
+        // operationName, and we include the region if the service is modeled with
+        // @sigv4.
+        String operation = executionAttributes.getAttribute(SdkExecutionAttribute.OPERATION_NAME);
+        // TODO: Do this only if sigv4
+        Region region = executionAttributes.getAttribute(AwsExecutionAttribute.AWS_REGION);
+
+        // Use the operation-level auth scheme resolver if the customer
+        // specified an override, otherwise fall back to the one on the client.
+        // TODO: The fallback should happen while setting in execution attributes
+        AcmAuthSchemeProvider authSchemeProvider =
+            (AcmAuthSchemeProvider) executionAttributes.getAttribute(SdkInternalExecutionAttribute.AUTH_SCHEME_RESOLVER);
+
+        return authSchemeProvider.resolveAuthScheme(AcmAuthSchemeParams.builder()
+                                                                       .operation(operation)
+                                                                       .region(region)
+                                                                       .build());
+    }
+
+    /**
+     * From a list of possible auth options for this request, determine which auth scheme should be used.
+     */
+    private SelectedAuthScheme<?> selectAuthScheme(List<AuthSchemeOption> authOptions, ExecutionAttributes executionAttributes) {
+
+        // TODO: This should be "merged" earlier, with request preferred over client
+        Map<String, AuthScheme<?>> authSchemes =
+            executionAttributes.getAttribute(SdkInternalExecutionAttribute.AUTH_SCHEMES);
+
+        IdentityProviderConfiguration identityResolvers =
+            executionAttributes.getAttribute(SdkInternalExecutionAttribute.IDENTITY_PROVIDER_CONFIGURATION);
+
+        StringBuilder failureReasons = new StringBuilder();
+
+        // Check each option, in the order the auth scheme resolver proposed them.
+        for (AuthSchemeOption authOption : authOptions) {
+            // If we're using no-auth, don't consider which options are enabled.
+            if (authOption.schemeId().equals("smithy.auth#noAuth")) {
+                return new SelectedAuthScheme(null, null, authOption);
```

Comment:
This bit does not work properly, above we require the params to be not null:

```
        this.identityProvider = Validate.paramNotNull(identityProvider, "identityProvider");
        this.signer = Validate.paramNotNull(signer, "signer");
```

Reply:
yes, this was fixed in https://github.com/aws/aws-sdk-java-v2/pull/4182#discussion_r1267367040